clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/VeniceAdaptiveBlobTransferTrafficThrottler.java

```diff
@@ -0,0 +1,171 @@
+  private static final Logger LOGGER = LogManager.getLogger(VeniceAdaptiveBlobTransferTrafficThrottler.class);
+  private static final String THROTTLER_NAME_SUFFIX = "TrafficBlobTransferThrottler";
+  private final String throttlerName;
+  private final List<BooleanSupplier> limiterSuppliers = new ArrayList<>();
+  private final List<BooleanSupplier> boosterSuppliers = new ArrayList<>();
+  private final AtomicReference<GlobalChannelTrafficShapingHandler> globalChannelTrafficShapingHandler =
+      new AtomicReference<>();
+  private final long baseRate;
+  private final int signalIdleThreshold;
+  private int currentFactor = 100;
+  private static final int MIN_FACTOR = 20;
+  private static final int MAX_FACTOR = 200;
+  private final int rateUpdatePercentage;
+  private final boolean isWriteThrottler;
+  private int signalIdleCount;
+
+  public VeniceAdaptiveBlobTransferTrafficThrottler(
+      int singleIdleThreshold,
+      long baseRate,
+      int rateUpdatePercentage,
+      boolean isWriteThrottler) {
+    this.throttlerName = (isWriteThrottler ? "Write" : "Read") + THROTTLER_NAME_SUFFIX;
+    this.baseRate = baseRate;
+    this.signalIdleThreshold = singleIdleThreshold;
+    this.isWriteThrottler = isWriteThrottler;
+    this.rateUpdatePercentage = rateUpdatePercentage;
+    LOGGER.info(
+        "Created adaptive throttler with name: {}, base rate: {}, change delta: {}",
+        throttlerName,
+        baseRate,
+        rateUpdatePercentage);
+  }
+
+  @Override
+  public void registerLimiterSignal(BooleanSupplier supplier) {
+    limiterSuppliers.add(supplier);
+  }
+
+  @Override
+  public void registerBoosterSignal(BooleanSupplier supplier) {
+    boosterSuppliers.add(supplier);
+  }
+
+  @Override
+  public void checkSignalAndAdjustThrottler() {
+    if (globalChannelTrafficShapingHandler.get() == null) {
+      LOGGER.info("Related traffic shaping handler does not exist, will not update the throttling number");
+      return;
+    }
+    boolean isSignalIdle = true;
+    boolean hasLimiterRate = false;
+    for (BooleanSupplier supplier: limiterSuppliers) {
+      if (supplier.getAsBoolean()) {
+        hasLimiterRate = true;
+        isSignalIdle = false;
+        signalIdleCount = 0;
+      }
+    }
+    // If any limiter signal is true do not booster the throttler
+    if (hasLimiterRate) {
+      if (currentFactor > MIN_FACTOR) {
+        currentFactor -= rateUpdatePercentage;
+        if (currentFactor < MIN_FACTOR) {
+          currentFactor = MIN_FACTOR;
+        }
+        updateThrottlerNumber();
+        LOGGER.info(
+            "Found limiter signal for {}, adjusting throttler factor to: {} with throttle rate: {}",
+            throttlerName,
+            currentFactor,
+            currentFactor * baseRate / 100);
+      }
+      return;
+    }
+    boolean hasBoosterRate = false;
+    for (BooleanSupplier supplier: boosterSuppliers) {
```

Comment:
Quick question.
Shall we boost only if all the booster signals are true?
For example.
```
Limiter signals:
read latency > 10ms
write latency > 5 mins

Booster signals:
read latency < 2 ms
write latency < 1 min
```
In the above example, shall we boost the rate if all the booster signals are true?
For example, the situation could be:
```
read latency < 2ms
write latency ~= 4.9 min
```